ports/esp8266/modneopixel.c

```diff
@@ -0,0 +1,200 @@
+        if (self->bpp<4) { continue; }
+        self->buf[offset++] = y;
+    }
+
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(NeoPixel_fill_obj, NeoPixel_fill);
+
+STATIC mp_obj_t NeoPixel_write(mp_obj_t self_in) {
+    NeoPixel_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    esp_neopixel_write(self->pin->phys_port, self->buf, self->n*self->bpp, true);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_1(NeoPixel_write_obj, NeoPixel_write);
+
+STATIC mp_obj_t NeoPixel_subscr(mp_obj_t self_in, mp_obj_t index_in, mp_obj_t val_in) {
+    NeoPixel_obj_t* self = MP_OBJ_TO_PTR(self_in);
+    size_t index = mp_obj_get_int(index_in);
+    if (index >= self->n) {
+        nlr_raise(mp_obj_new_exception(&mp_type_IndexError));
+    }
+
+    if (val_in == MP_OBJ_NULL) {
+        // delete
+        nlr_raise(mp_obj_new_exception(&mp_type_NotImplementedError));
+    } else if (val_in == MP_OBJ_SENTINEL) {
+        // read
+        size_t offset = self->bpp * index;
+        mp_obj_t tuple[4];
+        tuple[1] = mp_obj_new_int(self->buf[offset++]); // g
+        tuple[0] = mp_obj_new_int(self->buf[offset++]); // r
+        tuple[2] = mp_obj_new_int(self->buf[offset++]); // b
+        if (self->bpp > 3) {
+            tuple[3] = mp_obj_new_int(self->buf[offset++]); // y
+        }
+        return mp_obj_new_tuple(self->bpp, tuple);
+    } else {
+        // write
+        size_t val_len;
+        mp_obj_t *val;
+        mp_obj_get_array(val_in, &val_len, &val);
+        NeoPixel_assert_color(self, val_len);
+
+        size_t offset = self->bpp * index;
+        self->buf[offset++] = mp_obj_get_int(val[1]); // g
+        self->buf[offset++] = mp_obj_get_int(val[0]); // r
+        self->buf[offset++] = mp_obj_get_int(val[2]); // b
+        if (self->bpp > 3) {
+            self->buf[offset++] = mp_obj_get_int(val[3]); // y
+        }
+        return mp_const_none;
+    }
+}
+
+STATIC void NeoPixel_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+    NeoPixel_obj_t* self = MP_OBJ_TO_PTR(self_in);
+
+    if (dest[0] != MP_OBJ_NULL) {
+        // delete/store attribute
+        nlr_raise(mp_obj_new_exception(&mp_type_NotImplementedError));
+    }
+
+    if (attr == MP_QSTR_n) {
+        dest[0] = mp_obj_new_int(self->n);
+    } else if (attr == MP_QSTR_bpp) {
+        dest[0] = mp_obj_new_int(self->bpp);
+    } else if (attr == MP_QSTR_write) {
+        dest[0] = MP_OBJ_FROM_PTR(MP_ROM_PTR(&NeoPixel_write_obj));
+        dest[1] = self_in;
+    } else if (attr == MP_QSTR_fill) {
+        dest[0] = MP_OBJ_FROM_PTR(MP_ROM_PTR(&NeoPixel_fill_obj));
+        dest[1] = self_in;
+    } else {
+        nlr_raise(mp_obj_new_exception(&mp_type_AttributeError));
+    }
+}
+
+STATIC mp_obj_t NeoPixel___del__(mp_obj_t self_in) {
+    NeoPixel_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    m_del(uint8_t, self->buf, self->buf_size);
+    self->buf = NULL;
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_1(NeoPixel___del__obj, NeoPixel___del__);
```

Comment:
You don't really need this finaliser, memory will be reclaimed by the GC anyway during the same collection phase that would have called this finaliser.